db/db_impl/db_impl_compaction_flush.cc

```diff
@@ -1824,70 +1824,73 @@ Status DBImpl::AtomicFlushMemTables(
 // Called should check status and flush_needed to see if flush already happened.
 Status DBImpl::WaitUntilFlushWouldNotStallWrites(ColumnFamilyData* cfd,
                                                  bool* flush_needed) {
-  {
-    *flush_needed = true;
-    InstrumentedMutexLock l(&mutex_);
-    uint64_t orig_active_memtable_id = cfd->mem()->GetID();
-    WriteStallCondition write_stall_condition = WriteStallCondition::kNormal;
-    do {
-      if (write_stall_condition != WriteStallCondition::kNormal) {
-        // Same error handling as user writes: Don't wait if there's a
-        // background error, even if it's a soft error. We might wait here
-        // indefinitely as the pending flushes/compactions may never finish
-        // successfully, resulting in the stall condition lasting indefinitely
-        if (error_handler_.IsBGWorkStopped()) {
-          return error_handler_.GetBGError();
-        }
-
-        TEST_SYNC_POINT("DBImpl::WaitUntilFlushWouldNotStallWrites:StallWait");
-        ROCKS_LOG_INFO(immutable_db_options_.info_log,
-                       "[%s] WaitUntilFlushWouldNotStallWrites"
-                       " waiting on stall conditions to clear",
-                       cfd->GetName().c_str());
-        bg_cv_.Wait();
-      }
-      if (cfd->IsDropped()) {
-        return Status::ColumnFamilyDropped();
-      }
-      if (shutting_down_.load(std::memory_order_acquire)) {
-        return Status::ShutdownInProgress();
-      }
-
-      uint64_t earliest_memtable_id =
-          std::min(cfd->mem()->GetID(), cfd->imm()->GetEarliestMemTableID());
-      if (earliest_memtable_id > orig_active_memtable_id) {
-        // We waited so long that the memtable we were originally waiting on was
-        // flushed.
-        *flush_needed = false;
-        return Status::OK();
-      }
-
-      const auto& mutable_cf_options = *cfd->GetLatestMutableCFOptions();
-      const auto* vstorage = cfd->current()->storage_info();
-
-      // Skip stalling check if we're below auto-flush and auto-compaction
-      // triggers. If it stalled in these conditions, that'd mean the stall
-      // triggers are so low that stalling is needed for any background work. In
-      // that case we shouldn't wait since background work won't be scheduled.
-      if (cfd->imm()->NumNotFlushed() <
-              cfd->ioptions()->min_write_buffer_number_to_merge &&
-          vstorage->l0_delay_trigger_count() <
-              mutable_cf_options.level0_file_num_compaction_trigger) {
-        break;
-      }
+  *flush_needed = true;
+  InstrumentedMutexLock l(&mutex_);
+  uint64_t orig_active_memtable_id = cfd->mem()->GetID();
+
+  auto check_flush_causes_stall =
+      [&]() -> std::pair<Status, WriteStallCondition> {
```

Comment:
Maybe I'm not used to read code with lambda yet but somehow I don't feel that this helps with readability. Can it be a actual function instead? A function in anonymous namespace or DBImpl?
As people have pointed out, Google C++ Style now suggests people try to break down functions longer than 40-lines, and maybe we can break down to two functions if possible.

Reply:
The goal of this PR was to disentangle the stall check from the wait loop. It is ok if it doesn't improve readability, it only took a few minutes to try, so I have no problem closing it (I think if we change the goal to targeting 40 line functions, then this function is not the right place to start :) ).